### PR TITLE
Apple Pay Marks

### PR DESCRIPTION
**Pull request by mnicpt**

### CODE REVIEW

File: src/funding/applepay/config.jsx [hunk 1]
```diff
@@ -25,7 +25,8 @@ export function getApplePayConfig() : FundingSourceConfig {
             BUTTON_LAYOUT.VERTICAL
         ],
 
-        Logo: ({ logoColor, optional }) => ApplePayLogo({ logoColor, optional }),
+        Logo:  ({ logoColor, optional }) => ApplePayLogo({ logoColor, optional }),
+        Marks: ({ ...props }) => ApplePayMark({ ...props }),
```
Comment (bluepnume) [inline src/funding/applepay/config.jsx:29]: Can we just call this `Mark` since there's only one?

File: src/marks/template.jsx [hunk 2]
```diff
@@ -88,6 +97,10 @@ export function MarksElement({ fundingEligibility, fundingSources, height, exper
                     .paypal-button-card:last-child {
                         margin-right: 0px;
                     }
+
+                    .paypal-mark .paypal-logo-applepay {
```
Comment (bluepnume) [inline src/marks/template.jsx:101]: Can we keep this in the applepay.js file, maybe in its own style tag?

File: src/ui/buttons/props.js [hunk 3]
```diff
@@ -190,6 +190,7 @@ export type RenderButtonProps = {|
     experiment : Experiment,
     vault : boolean,
     userIDToken : ?string,
+    applePay : Object,
```
Comment (bluepnume) [inline src/ui/buttons/props.js:193]: Can we use a full type for applePay? `Object` is very similar to using `any` and doesn't provide much type safety for inner keys.

File: src/zoid/buttons/util.js [hunk 4]
```diff
@@ -74,3 +74,35 @@ export function getVenmoExperiment(experiment : ?Experiment) : VenmoExperiment {
...
+                        'shippingmethodselected',
+                        'shippingcontactselected',
+                        'paymentauthorized',
+                        'cancel'
+                    ];
+                    if (validNames.indexOf(name) === -1) {
+                        // eslint-disable-next-line no-console
+                        console.error(`Invalid ApplePaySession event ${ name }`);
+                    }
+                    session[`on${ name }`] = handler;
```
Comment (bluepnume) [inline src/zoid/buttons/util.js:101]: Does applepay not support `addEventListener`?

**Comment by bluepnume** [comment]
Looks great to me! Just a few minor comments.

File: src/funding/applepay/config.jsx (hunk 1 above)
Comment (mstuart) [reply src/funding/applepay/config.jsx:29]: No, there's two.

<---

Comment (bluepnume) [reply src/funding/applepay/config.jsx:29]: onlyone

File: src/zoid/buttons/util.js (hunk 4 above)
Comment (mnicpt) [reply src/zoid/buttons/util.js:101]: Just thought it was a bit long.

File: src/zoid/buttons/util.js [hunk 5]
```diff
@@ -74,3 +74,35 @@ export function getVenmoExperiment(experiment : ?Experiment) : VenmoExperiment {
         enableVenmo: Boolean(isExperimentEnabled || isEnableFundingVenmo)
     };
 }
+
+export function applePaySession() : Function {
+    try {
+        if (!window.ApplePaySession) {
+            return undefined;
```
Comment (gregjopa) [inline src/zoid/buttons/util.js:81]: I think we can just use `return;` and that will return undefined.

```suggestion
            return;
```

**Comment by gregjopa** [comment]
PR looks good to me. CI is failing though with the following error:
```
Cannot create `MarkLogo` element because property `env` is missing in object type [1] but exists in props [2].
[incompatible-use]

   src/marks/template.jsx:34:20
   34|                 ? <MarkLogo
                          ^^^^^^^^

References:
   src/funding/common.jsx:84:14
   84|     Mark? : ({||}) => ChildType,
                    ^^^^ [1]
   src/marks/template.jsx:34:19
                         v--------
   34|                 ? <MarkLogo
   35|                     fundingEligibility={ fundingEligibility }
   36|                     locale={ getLocale() }
   37|                     experiment={ experiment }
   38|                     env={ env }
   39|                 />
                       -^ [2]
```

File: src/zoid/buttons/util.js [hunk 6]
```diff
@@ -74,3 +74,35 @@ export function getVenmoExperiment(experiment : ?Experiment) : VenmoExperiment {
...
+                    if (validNames.indexOf(name) === -1) {
+                        // eslint-disable-next-line no-console
+                        console.error(`Invalid ApplePaySession event ${ name }`);
+                    }
+                    session[`on${ name }`] = handler;
+                }
+            };
+        };
+    } catch (e) {
+        return undefined;
```
[the remark this answers is not in the document]
Comment (gregjopa) [reply src/zoid/buttons/util.js:109]: One more `undefined` to remove.

**Review by gregjopa — APPROVED** [review]
Looks good to me! 

I'm pretty new to Apple Pay to it would be good to get @bluepnume's 👍  as well.

**Comment by mnicpt** [comment]
Thanks @gregjopa

File: src/zoid/buttons/util.js [hunk 7]
```diff
@@ -74,3 +74,38 @@ export function getVenmoExperiment(experiment : ?Experiment) : VenmoExperiment {
...
+        }
+
+        return (version, request) => {
+            const session = new window.ApplePaySession(version, request);
+            return {
+                addEventListener: (name, handler) => {
+                    session.addEventListener(name, handler);
+                },
+                completeMerchantValidation: (validatedSession) => {
+                    session.completeMerchantValidation(validatedSession);
```
Comment (bluepnume) [inline src/zoid/buttons/util.js:91]: Do these return promises? Should we return those so the caller can await them?

Comment (mnicpt) [reply src/zoid/buttons/util.js:91]: @bluepnume No.  They are fire and forget and return void: https://developer.apple.com/documentation/apple_pay_on_the_web/applepaysession/1778012-completepayment .  We do have 30s to fire from getting the `onpaymentauthorized` event or they will invalidate the payment which is good if something happens when calling onApprove.

**Review by bluepnume — APPROVED** [review]

**Pull request merged**